cloudinit/distros/package_management/apt.py

```diff
@@ -0,0 +1,227 @@
+        command = ["eatmydata"]
+    else:
+        enabled = cfg.get("enabled")
+        command = cfg.get("command")
+
+        if isinstance(command, str):
+            command = [command]
+        elif not isinstance(command, list):
+            raise TypeError("apt_wrapper command must be a string or list")
+
+    if util.is_true(enabled) or (
+        str(enabled).lower() == "auto" and command and subp.which(command[0])
+    ):
+        return cast(List[str], command)
+    else:
+        return []
+
+
+class Apt(PackageManager):
+    name = "apt"
+
+    def __init__(
+        self,
+        runner: helpers.Runners,
+        *,
+        apt_get_wrapper_command: Sequence[str] = (),
+        apt_get_command: Optional[Sequence[str]] = None,
+        apt_get_upgrade_subcommand: Optional[str] = None,
+    ):
+        super().__init__(runner)
+        if apt_get_command is None:
+            apt_get_command = APT_GET_COMMAND
+        if apt_get_upgrade_subcommand is None:
+            apt_get_upgrade_subcommand = "dist-upgrade"
+        self.apt_command = tuple(apt_get_wrapper_command) + tuple(
+            apt_get_command
+        )
+
+        self.apt_get_upgrade_subcommand = apt_get_upgrade_subcommand
+        self.environment = os.environ.copy()
+        self.environment["DEBIAN_FRONTEND"] = "noninteractive"
+
+    @classmethod
+    def from_config(cls, runner: helpers.Runners, cfg: Mapping) -> "Apt":
+        return Apt(
+            runner,
+            apt_get_wrapper_command=get_apt_wrapper(
+                cfg.get("apt_get_wrapper")
+            ),
+            apt_get_command=cfg.get("apt_get_command"),
+            apt_get_upgrade_subcommand=cfg.get("apt_get_upgrade_subcommand"),
+        )
+
+    def update_package_sources(self):
+        self.runner.run(
+            "update-sources",
+            self.run_package_command,
+            ["update"],
+            freq=PER_INSTANCE,
+        )
+
+    @functools.lru_cache(maxsize=1)
+    def get_all_packages(self):
+        resp: str = subp.subp(["apt-cache", "pkgnames"]).stdout
+
+        # Searching the string directly and searching a list are both
+        # linear searches. Converting to a set takes some extra up front
+        # time, but resulting searches become binary searches and are much
+        # faster
+        return set(resp.splitlines())
+
+    def get_unavailable_packages(self, pkglist: Iterable[str]):
+        return [pkg for pkg in pkglist if pkg not in self.get_all_packages()]
+
+    def install_packages(self, pkglist: Iterable[str]) -> UninstalledPackages:
+        self.update_package_sources()
+        pkglist = util.expand_package_list("%s=%s", list(pkglist))
+        unavailable = self.get_unavailable_packages(
+            [x.split("=")[0] for x in pkglist]
+        )
+        LOG.debug(
+            "The following packages were not found by APT so APT will "
+            "not attempt to install them: %s",
+            unavailable,
+        )
```

Comment:
> This changes default behavior in the case where we had packages: [pkg1, pkg2] in that we now pay the cost of get_unavailable_packages for any previous #cloud-config which provides a generic pkg list.

This doesn't actually change **behavior** though, right? It just adds some search time?

> On distributions like debian where only one package_manager is defined, do we want to avoid this optional unavailable check and just attempt to install all packages provided and "fail hard" if some don't exist?

Our inheritance rules make this a little tricky. I'm also not sure I see a super compelling reason to.

> So, maybe that small cost is worth the better handling and support of snaps on stable releases.

This was my thinking. If you're installing packages, you're already adding more than a handful of seconds to your boot, so this should be fairly trivial compared to that. It doesn't impact the "fast boot" use case. Also, package install also comes after ssh, so it shouldn't impact time to ssh.

Reply:
> This doesn't actually change behavior though, right? It just adds some search time?
 Correct, yes no config changes or artifacts from cloud-init other than potential upfront cost to parse the apt-cache. Which again is not bad relative to the actual apt update we just performed and the software we need to install and configure during first boot anyway.... We've also added more error handling/coping if we want to fallback to snap'd packages in cases where the deb isn't available in apt, so I think it's a defensible cost to pay here.

>  Also, package install also comes after ssh, so it shouldn't impact time to ssh.
+1 to this too